sys/net/application_layer/sock_dodtls/sock_dodtls.c

```diff
@@ -0,0 +1,285 @@
+    if (res < 0 && res != CREDMAN_EXIST) {
+        DEBUG("Unable to add credential to credman\n");
+        _close_session(creds->tag, creds->type);
+        switch (res) {
+            case CREDMAN_NO_SPACE:
+                res = -ENOSPC;
+                break;
+            case CREDMAN_ERROR:
+            case CREDMAN_INVALID:
+            case CREDMAN_TYPE_UNKNOWN:
+            default:
+                res = -EINVAL;
+                break;
+        }
+        goto exit;
+    }
+    if ((res = sock_dtls_create(&_dtls_sock, &_udp_sock, creds->tag,
+                                SOCK_DTLS_1_2, SOCK_DTLS_CLIENT)) < 0) {
+        puts("Unable to create DTLS sock\n");
+        _close_session(creds->tag, creds->type);
+        goto exit;
+    }
+
+    start = _now_ms();
+    try_start = start;
+    while (((try_start = _now_ms()) - start) < SOCK_DODTLS_SESSION_TIMEOUT_MS) {
+        memset(&_server_session, 0, sizeof(_server_session));
+        if ((res = sock_dtls_session_init(&_dtls_sock, server,
+                                          &_server_session)) >= 0) {
+            uint32_t try_duration;
+
+            res = sock_dtls_recv(&_dtls_sock, &_server_session, _dns_buf,
+                                 sizeof(_dns_buf), timeout * US_PER_MS);
+            if (res == -SOCK_DTLS_HANDSHAKE) {
+                break;
+            }
+            DEBUG("Unable to establish DTLS handshake: %d (timeout: %luus)\n",
+                  -res, (long unsigned)timeout * US_PER_MS);
+            sock_dtls_session_destroy(&_dtls_sock, &_server_session);
+            try_duration = _now_ms() - try_start;
+            if (try_duration < timeout) {
+                _sleep_ms(timeout - try_duration);
+            }
+            /* see https://datatracker.ietf.org/doc/html/rfc6347#section-4.2.4.1 */
+            timeout *= 2U;
+        }
+        else {
+            DEBUG("Unable to initialize DTLS session: %d\n", -res);
+            sock_dtls_session_destroy(&_dtls_sock, &_server_session);
+        }
+    }
+    if (res != -SOCK_DTLS_HANDSHAKE) {
+        res = -ETIMEDOUT;
+        _close_session(creds->tag, creds->type);
+        goto exit;
+    }
+    else {
+        res = 0;
+    }
+
+    _cred_type = creds->type;
+    _cred_tag = creds->tag;
+    _id = (uint16_t)(random_uint32() & 0xffff);
+exit:
+    memset(_dns_buf, 0, sizeof(_dns_buf));  /* flush-out unencrypted data */
+    mutex_unlock(&_server_mutex);
+    return (res > 0) ? 0 : res;
+}
+
+static int _disconnect_server(void)
+{
+    int res = 0;
+
+    mutex_lock(&_server_mutex);
+    if (!_server_set()) {
+        goto exit;
+    }
+    _close_session(_cred_tag, _cred_type);
+    _cred_tag = CREDMAN_TAG_EMPTY;
+    _cred_type = CREDMAN_TYPE_EMPTY;
+exit:
+    mutex_unlock(&_server_mutex);
+    return res;
+}
+
+static uint32_t _now_ms(void)
+{
+    return ztimer_now(ZTIMER_MSEC);
+}
+
+static void _sleep_ms(uint32_t delay)
+{
+    ztimer_sleep(ZTIMER_MSEC, delay);
+}
```

Comment:
Are those wrapper functions really needed?

Reply:
They used to be, when `xtimer` was still supported. Let's keep the in case of YATU (yet another timer update ;-)).